micro-rdk/src/common/conn/viam.rs

```diff
@@ -0,0 +1,756 @@
+                        let mut api = WebRtcApi::new(
+                            self.executor.clone(),
+                            sig.0,
+                            sig.1,
+                            conf.cert.clone(),
+                            ip,
+                            conf.dtls.make().unwrap(),
+                        );
+                        let sdp = api.answer(0).await.unwrap();
+                        let mut c = WebRTCConnection {
+                            webrtc_api: api,
+                            sdp: sdp.0,
+                            server: None,
+                            robot: self.robot.clone(),
+                            prio: sdp.1,
+                        };
+                        let p = c.open_data_channel().await.unwrap();
+                        let p = c.run().await.unwrap();
+                    }
+                }
+            }
+        }
+    }
+    async fn server_peer_http2(&self, io: Box<dyn HTTP2Stream>) {
+        let robot = self.robot.clone();
+        let exec = self.executor.clone();
+        let srv = GrpcServer::new(robot, GrpcBody::new());
+        let srv = Box::new(
+            http2::Builder::new(exec)
+                .initial_connection_window_size(2048)
+                .initial_stream_window_size(2048)
+                .max_send_buf_size(4096)
+                .max_concurrent_streams(2)
+                .serve_connection(io, srv),
+        );
+        let t = self.executor.spawn(srv).await;
+    }
+}
+
+struct SignalingTask {
+    sender: Sender<AppSignaling>,
+    rpc_host: String,
+}
+
+impl PeriodicAppClientTask for SignalingTask {
+    fn get_default_period(&self) -> Duration {
+        Duration::from_secs(0)
+    }
+    fn name(&self) -> &str {
+        "SignalingTask"
+    }
+    fn invoke<'b, 'a: 'b>(
+        &'a self,
+        app_client: &'b AppClient,
+    ) -> std::pin::Pin<Box<dyn Future<Output = Result<Option<Duration>, AppClientError>> + 'b>>
+    {
+        Box::pin(async {
+            let sig_pair = app_client.initiate_signaling(self.rpc_host.clone()).await?;
+            let _ = self.sender.send(sig_pair).await; // TODO deal with result, sending on a close channel will never succeed. The limit here is that SignalingTask will be allocated for the lifetime of the ViamServer.
+            Ok(None)
+        })
+    }
+}
+
+struct AppClientTaskRunner<'a> {
+    task: Option<
+        std::pin::Pin<Box<dyn Future<Output = Result<Option<Duration>, AppClientError>> + 'a>>,
+    >,
+    invoker: &'a Box<dyn PeriodicAppClientTask>, //need to impl deref?
+    app_client: &'a AppClient,
+    timer: Option<async_io::Timer>,
+}
+
+impl<'a> Future for AppClientTaskRunner<'a> {
+    type Output = Result<(), AppClientError>;
+    fn poll(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        if let Some(&mut ref mut timer) = self.timer.as_mut() {
```

Comment:
Could we make use of an enum to simplify this? We can set task as an instance of the enum and switch whether the enum is the timer or the actual task based on what the task was previously and whether it returns `Poll::Ready` or not

Reply:
I updated the code hopefully is a bit easier to follow through